### DISCUSSION TITLE
v0.23.0-rc5: google oauth Realtime connection closed (cancelled request)

### DISCUSSION THREAD

**Discussion by Plorenzo** [pocketbase/pocketbase#5662] [Q&A] [answered]
I'm trying to test Google oauth integration. I created my app in gconsole and got my id and secret.

I have a button that triggers the sign in:
```ts
async function signInWithGoogle() {
		try {
			const authData = await pb.collection('users').authWithOAuth2({ provider: 'google' });
			console.debug('signed in with google', authData);
		} catch (error) {
			console.error(error);
		}
	} 
```

When I trigger the sign in, everything seems to be working fine, a new window pops up asking to log in with Google credentials, the oauth consent screen appears and all the log in flow goes as usual. The issue is when the flow finishes and goes back to the app, there is no new user created. 

This is the only log I can find `DEBUG Realtime connection closed (cancelled request)`
```
INFO GET /api/collections/users/auth-methods?fields=mfa,otp,password,oauth2
DEBUG Realtime connection established.
└─ {"clientId":"REDACTED"}
DEBUG Realtime subscriptions updated.
└─ {"clientId":"REDACTED","subscriptions":["@oauth2"]}
INFO POST /api/realtime
INFO GET /api/oauth2-redirect?state= REDACTED&code= REDACTED&scope=email+profile+https://www.googleapis.com/auth/userinfo.email+openid+https://www.googleapis.com/auth/userinfo.profile&authuser=0&prompt=none
[0.00ms] SELECT `_externalAuths`.* FROM `_externalAuths` WHERE (NOT (`providerId`='')) AND (`collectionRef`='_pbc_344172009' AND `provider`='google' AND `providerId`='111302864989861237770') ORDER BY `created` DESC LIMIT 1
[0.00ms] SELECT `users`.* FROM `users` WHERE `email`='REDACTED@gmail.com' LIMIT 1
[0.00ms] SELECT `_superusers`.* FROM `_superusers` WHERE `_superusers`.`id`='fyf72c68nvlr665' LIMIT 1
INFO POST /api/collections/_superusers/auth-refresh
[0.00ms] SELECT `_superusers`.* FROM `_superusers` WHERE `_superusers`.`id`='fyf72c68nvlr665' LIMIT 1
INFO GET /api/settings
DEBUG Realtime connection closed (cancelled request)
└─ {"clientId":"hDOlznYhJrejez26nXJI1pfVLejfdeA11SDJvuxp"}
```

This is the redirect url im using `https://example.com/api/oauth2-redirect`

I realize that this may not have anything to do with PB, but as I'm not getting any google error in the login flow o in the console I figure it's better to ask. 

I tried with several browser, all present the same behaviour

**Comment by ganigeorgiev** [comment]
> The issue is when the flow finishes and goes back to the app, there is no new user created.

Are you sure, because the SQL logs suggest otherwise? If the user is not persisted then double check your custom app hooks for missing `e.Next()` calls or similar.

The "realtime connection closed" debug message is expected and there is nothing wrong with it. It indicates that the realtime connection was closed from the client-side and this is fine because that is the expected behavior when the "all-in-one" OAuth2 flow end.

With v0.23.0 we have automated tests for the OAuth2 and I've also recently tested the OAuth2 flow with the monday.com, Notion and Zoho OIDC providers and can confirm that it works correctly.

**Reply by ganigeorgiev** [reply]
Also I forgot to mention that if you were previously already authenticated (aka. `pb.authStore.token` is not empty before the OAuth2 call) note that the OAuth2 user will be linked to that authenticated user, if not linked to someone else, and will not create a new one, so check in the Admin UI for the existing users and their available "Authorized providers" entries.

**Comment by Plorenzo** [comment]
I'm sure the user is not created, not sure if it's my fault (: 

If the user is being created I should see an INSERT in the logs no? 

I have the following hook to add some required fields to the `user` record 

```go
app.OnRecordAuthWithOAuth2Request("users").BindFunc(func(e *core.RecordAuthWithOAuth2RequestEvent) error {

		app.Logger().Debug("OnRecordAuthWithOAuth2Request")
		e.Record.Set("role", "user")
		e.Record.Set("status", "active")

		return e.Next()
	})
```
I'm not sure if I need to `app.Save(e.Record)` inside the hook or as it's a before hook it will get saved in the next step. Either way I tried both and same result. 

I will continue debugging to see if it's an issue with the my hooks

**Reply by ganigeorgiev** [reply]
> If the user is being created I should see an INSERT in the logs no?

Not necessary. See the comment above -https://github.com/pocketbase/pocketbase/discussions/5662#discussioncomment-10927405.


> I'm not sure if I need to app.Save(e.Record) inside the hook...

In the above hook, `e.Record` may not be available (aka. nil) so you need to check for this too before doing any record operations. You need to manually call `Save()` if it is an existing record (aka. `e.Record != nil`) since there is no guarantee that the OAuth2 operation will perform update to save your changes (sometimes it does, for example updating the verified state of the record if not already).

**Reply by ganigeorgiev** [reply]
Again, check your existing users and whether they have "Authorized providers" entry (inspect also the returned auth data record, aka. `console.log(authData.record.id)` and search for the user in the UI).

**Reply by ganigeorgiev** [reply, marked as answer]
Also note that if the desired behavior is to set default field values with the OAuth2 user request creation, then I recommend to attach to the `OnRecordCreateRequest` hook (the OAuth2 operation will trigger it).

**Reply by ganigeorgiev** [reply]
I've just tested it locally and it works correctly for me. I'm pretty sure that in your case the OAuth2 account is linked to some of your existing users.

If that's not the case feel free to create a minimal reproducible and I'll try to have a look at it.

**Reply by Plorenzo** [reply]
thanks again Gani! As suggested I change the hook to `OnRecordCreateRequest` and now works without issues.
